File: src/com/basmilius/ps/bastools/framework/all/completion/FunctionCallAllArgumentsCompletionProvider.java
package com.basmilius.ps.bastools.framework.all.completion;

import com.intellij.codeInsight.completion.*;
import com.intellij.codeInsight.lookup.LookupElementBuilder;
import com.intellij.codeInsight.lookup.LookupValueWithPriority;
import com.intellij.patterns.ElementPattern;
import com.intellij.patterns.PlatformPatterns;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.util.ProcessingContext;
import com.jetbrains.php.lang.parser.PhpElementTypes;
import com.jetbrains.php.lang.psi.elements.Function;
import com.jetbrains.php.lang.psi.elements.Parameter;
import org.jetbrains.annotations.NotNull;

public class FunctionCallAllArgumentsCompletionProvider extends CompletionProvider<CompletionParameters>
{

	@Override
	protected void addCompletions (@NotNull final CompletionParameters parameters, final ProcessingContext context, @NotNull final CompletionResultSet results)
	{
		final Function func = PsiTreeUtil.getParentOfType(parameters.getOriginalPosition(), Function.class);

		if (func != null)
		{
			final Parameter[] ps = func.getParameters();
			final String[] pss = new String[ps.length];

			for (int i = 0; i < ps.length; i++)
			{
				pss[i] = "$" + ps[i].getName();
			}

			results.addElement(PrioritizedLookupElement.withPriority(LookupElementBuilder.create(String.join(", ", pss)), LookupValueWithPriority.HIGHER));
		}
	}

	public static ElementPattern PLACE = PlatformPatterns.psiElement().inside(PlatformPatterns.psiElement(PhpElementTypes.FUNCTION_CALL));

}

File: src/com/basmilius/ps/bastools/framework/all/completion/FunctionCallAllArgumentsCompletionProvider.kt
package com.basmilius.ps.bastools.framework.all.completion

import com.intellij.codeInsight.completion.CompletionParameters
import com.intellij.codeInsight.completion.CompletionProvider
import com.intellij.codeInsight.completion.CompletionResultSet
import com.intellij.codeInsight.completion.PrioritizedLookupElement
import com.intellij.codeInsight.lookup.LookupElementBuilder
import com.intellij.codeInsight.lookup.LookupValueWithPriority
import com.intellij.patterns.ElementPattern
import com.intellij.patterns.PlatformPatterns
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.util.ProcessingContext
import com.jetbrains.php.lang.parser.PhpElementTypes
import com.jetbrains.php.lang.psi.elements.Function

/**
 * Class FunctionCallAllArgumentsCompletionProvider
 *
 * @author Bas Milius
 * @package com.basmilius.ps.bastools.framework.all.completion
 */
class FunctionCallAllArgumentsCompletionProvider : CompletionProvider<CompletionParameters>()
{

	/**
	 * {@inheritdoc}
	 *
	 * @author Bas Milius
	 */
	override fun addCompletions(parameters : CompletionParameters, context : ProcessingContext, results : CompletionResultSet)
	{
		val func = PsiTreeUtil.getParentOfType(parameters.originalPosition, Function::class.java)

		if (func != null)
		{
			val ps = func.parameters
			val pss = arrayOfNulls<String>(ps.size)

			for (i in ps.indices)
				pss[i] = "$" + ps[i].name

			results.addElement(PrioritizedLookupElement.withPriority(LookupElementBuilder.create(pss.joinToString(", ")), LookupValueWithPriority.HIGHER.toDouble()))
		}
	}

	/**
	 * Companion Object for FunctionCallAllArgumentsCompletionProvider
	 *
	 * @author Bas Milius
	 * @package com.basmilius.ps.bastools.framework.all.completion
	 */
	companion object
	{

		/**
		 * Gets the PLACE where this completion should go.. (or something).
		 */
		var PLACE : ElementPattern<*> = PlatformPatterns.psiElement().inside(PlatformPatterns.psiElement(PhpElementTypes.FUNCTION_CALL))

	}

}
